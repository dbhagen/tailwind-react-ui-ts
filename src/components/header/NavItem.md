Individual menu items in navigation bar.
