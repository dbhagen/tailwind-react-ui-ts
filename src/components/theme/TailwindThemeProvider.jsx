import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
