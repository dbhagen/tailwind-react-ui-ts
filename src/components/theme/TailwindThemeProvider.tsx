import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

// TODO: Remove Any Types
const TailwindThemeProvider = ({ theme, children }: { theme: any, children: any }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
