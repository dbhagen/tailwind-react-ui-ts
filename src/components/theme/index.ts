export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
