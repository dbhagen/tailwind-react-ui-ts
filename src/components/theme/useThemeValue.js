import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
