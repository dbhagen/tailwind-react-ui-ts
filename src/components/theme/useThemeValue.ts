import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, values: Object, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(undefined, values, prefix)
}
