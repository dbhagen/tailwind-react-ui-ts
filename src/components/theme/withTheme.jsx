import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
