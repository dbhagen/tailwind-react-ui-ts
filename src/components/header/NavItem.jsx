import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Touchable } from '../primitives'

const NavItem = ({
  theme,
  is,
  children,
  header: { style, screen },
  active,
  ...rest
}) => {
  const textColor = style.text || theme.textColors.on.primary
  const bgColor = style.bg || theme.brandColors.primary

  const responsive = screen
    ? {
      [`m-${screen}`]: { t: 0, r: theme.spacing.sm }
    }
    : {}

  return (
    <Touchable
      is={is}
      focusable
      text={!active ? style.text : style.bg}
      bg={active ? textColor : undefined}
      bg-hocus={textColor}
      text-hocus={bgColor}
      p={{ x: theme.spacing.md, y: theme.spacing.sm }}
      m={{ t: theme.spacing.sm }}
      rounded={theme.radius}
      block
      aria-current={active ? 'page' : undefined}
      {...responsive}
      {...rest}
    >
      {children}
    </Touchable>
  )
}

NavItem.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string])
  }),
  active: PropTypes.bool
}

NavItem.defaultProps = {
  is: 'a',
  children: undefined,
  header: { style: {}, screen: 'lg' },
  active: false
}

export { NavItem as component }
export default withTheme(NavItem)
