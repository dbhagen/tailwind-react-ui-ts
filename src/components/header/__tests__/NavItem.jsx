import React from 'react'
import { shallow } from 'enzyme'

import { component as NavItem } from '../NavItem'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign(
    {
      theme: defaultTheme,
      header: { style: { bg: 'blue-400', text: 'white' } },
    },
    testProps,
  )

  const wrapper = shallow(<NavItem {...props}>Link</NavItem>)

  return {
    props,
    wrapper,
  }
}

describe('NavItem', () => {
  it('renders matching snapshot', () => {
    // eslint-disable-next-line
    const { wrapper } = setup({ is: props => <a {...props} href="#test" /> })

    expect(wrapper).toMatchSnapshot()
  })
})
