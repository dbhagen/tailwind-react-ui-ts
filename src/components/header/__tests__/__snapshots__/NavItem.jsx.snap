// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem renders matching snapshot 1`] = `
<Touchable
  bg-hocus="white"
  block={true}
  disabled={false}
  focusable={true}
  is={[Function]}
  m={
    Object {
      "t": 2,
    }
  }
  p={
    Object {
      "x": 4,
      "y": 2,
    }
  }
  rounded="rounded"
  text="white"
  text-hocus="blue-400"
>
  Link
</Touchable>
`;
